### ISSUE TITLE
clsparseSCsrMatrixfromFile() returns clsparseInvalidSize. 

### ISSUE THREAD

**Issue by kvaragan** [clMathLibraries/clSPARSE#144]
I am trying to run default benchmark function SpMdV on test file: Bell & Garland cant.mtx file.
I am getting error clsparseInvalidSize. on further debugging the actual error occurs at validateMemObject() due to current_size != required_size. This error was not present before, I don't know why this is occurring? I tried other bell & garland matrices the output is same.

Environment: W9100, Ubuntu 14.04 LTS.
The same behaviour is also observed on windows.
However the older version codes are working fine.

**Comment by jpola** [comment]
To make it work locally please comment lines from 660 from mm-reader.cpp:
Fix should looks like this:
```c++
         clsparseStatus validationStatus = clsparseSuccess;
        // = validateMemObject(pCsrMatx->values,
        //                              mm_reader.GetNumNonZeroes() * sizeof(cl_double));

        // I dont want to reallocate buffer because I suppress the users buffer memory flags;
        // It is users responsibility to provide good buffer;
        // if (validationStatus != clsparseSuccess)
        //    return validationStatus;

        validationStatus = validateMemObject(pCsrMatx->colIndices,
                                             mm_reader.GetNumNonZeroes() * sizeof(cl_int));
```
The issue is related to mm_reader. This object for reading mtx file is reading the file twice. The first pass is just to read the header of the file. From header it is taking the necessary number of rows, cols and nnz to be allocated. The second pass is reading the values and values locations. To optimize the code (reduce the acutal number of nnz) Kent was additionally testing the read values aganist zero. This lead to inconsistency between declared parameters from header and the actual number of nnz values.

@kknox I see several options
1. Do not check for val == 0 during the reading function
2. Check for val == 0 but not validate the matrix values buffer size

I personally vote for option nr 1. Because it is not our responsibility to properly creating the sparse matrix even if there are explicit zeros. We may also combine the 1 & 2 with some extra flag like bool check_for_extra_zeros.

In addition checking for explicit zeros may false some benchmarks when comparing to results from publications. I didn't see that anybody was checking for explicit zeros. In case of cant.mtx the difference between header and read number of zeros is 62451. I nuderstand that it may be waste of memory, but again it was the user who created that file.

**Comment by jlgreathouse** [comment]
Hi @jpola,

I wrote a fix for this problem while trying to debug Issue #141. Correct me if I'm wrong, but it seems that the validateMemObject function as written is too strict. On line 67 it is testing "if(current_size != required_size)" and returning clsparseInvalidSize (which is then silently causing failures in a lot of our benchmarks).

I haven't submitted a PR for this yet, but I fixed the problem seen by @kvaragan by changing it to current_size < required_size. See https://github.com/jlgreathouse/clSPARSE/commit/39c40cf2a871d90215f9d12f11d4028cbff17735. I believe this should be OK, because the validate function should ensure that whatever we want to put in the buffer should fit. I don't think strict size equality is required.

**Comment by kvaragan** [comment]
Hi @jpola, I agree nr 1 is not needed, I am also surprised why sparse storage format inputs are having some elements with val == 0. Maybe for very large MTX files we can use this technique for fear of not enough memory but otherwise I feel its not needed.

**Comment by jlgreathouse** [comment]
I also agree with at least including the option to save explicit zero values into the matrix when reading it in. All of the academic benchmarks we compare against save the explicit zero values, as does our internal code we used for the papers we wrote about CSR-Adaptive.

The [UFL collection claims](https://www.cise.ufl.edu/research/sparse/matrices/legend.html) that keeping these around is useful because "they can represent important information, such as entries that will become nonzero later on in a subsequent interation of a nonlinear solver".

**Comment by jpola** [comment]
@jlgreathouse, unfortunatelly I can't agree on your suggestion regarding current_size < required_size. If the buffer will be smaller than required the algorithms will have undefined behaviour.

**Comment by jlgreathouse** [comment]
Err, you're right. Sorry, I ended up doing exactly the opposite of what I'd hoped to do there. The general point still stands, though: should we not compare "if(current_size >= required_size)" instead of != ?

That, of course, doesn't fix the problem being discussed here but may prevent issues in the future.

**Comment by jpola** [comment]
I was thinking about the same. We should allow users to allocate larger buffers than required. They i.e. may store several vectors in one buffer and play with the offsets (even if they are not fully supported right now). 

@kknox 
This function is your child :) How do you wish this issue to be resolved?

**Comment by kknox** [comment]
@jpola To give credit where it's properly due, the mm-reader code was originally written by one Chip Freitag; I've ported and refactored his code for clsparse use.

On to the question: I agree the 'explicit' zeroes setting should be configurable, and we should probably change the default to include explicit zeroes.  We could easily implement this with flags that are passed into the API to control mm-reader behavior.  

As for the problems with validateMemObject, I think Jakub and Joseph are narrowing down what the appropriate behavior should be.  There should be no harm in allocating buffers bigger than they need to be, and we want to help catch user error only if we minimize false positives.  

As another general comment, I also believe that sanity checks like these should only be done in DEBUG builds, and RELEASE builds should minimize all calls into the OpenCL runtime layer.

**Comment by jpola** [comment]
I will create two separate tickets. First for sanity checks, second for mm-reader.

**Comment by kknox** [comment]
Do these two separate tickets supersede this one?  Is this now closed?  Or is this ticket still tracking a unique problem?  I'm wondering about the code  in @jlgreathouse fork that will eventually be merged.

**Comment by jlgreathouse** [comment]
I sat down to hammer on this a bit, and here are some updates:
@jpola, I've convinced myself that the patch I linked (https://github.com/jlgreathouse/clSPARSE/commit/39c40cf2a871d90215f9d12f11d4028cbff17735) is correct. Note that "if(current_size < required_size)" is the **failure** case. If the current size of your buffer is not large enough to hold the required number of NNZs, return clSparseInvalidSize.

The current_size != required_size case happens for two reasons where we allocate more space than we actually need:
* As mentioned above, we currently do not read in explicit zero values from matrix market files. As such, the NNZ listed in the MM header is larger than the NNZ we actually put into the data structure
   * @kknox, I don't think this one is Chip's fault -- someone added in 'const int exp_zeroes = 0;'. It used to come from a command line parameter. :)
* When reading in a symmetric matrix, we take the NNZ in the matrix market header and multiply it by two in order to decide how much space to allocate. In actuality, a symmetric matrix will have less than 2x the number of non-zeros, because values on the diagonal are not stored twice.

As such, fixing only Issue #146 will not solve this current issue. I believe that we must have the patch https://github.com/jlgreathouse/clSPARSE/commit/39c40cf2a871d90215f9d12f11d4028cbff17735 and allow slight over-allocations of space in these cases.

Thoughts?
